Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -69,46 +173,78 @@ class CDMPrivateOpenCDM : public CDMPrivate {
-    Ref<SharedBuffer> message() const { ASSERT(m_message); return Ref<SharedBuffer>(*m_message.get()); }
+    using ChallengeGeneratedCallback = Function<void(Session*)>;
+    using KeyUpdatedCallback = Function<void(Session*, OCDMKeyStatus, RefPtr<SharedBuffer>)>;
+    using SessionChangedCallback = Function<void(Session*, bool success, RefPtr<SharedBuffer>)>;
+
+    static Ref<Session> create(OpenCDMAccessor&, const String&, const char*, Ref<WebCore::SharedBuffer>&&, LicenseType, Ref<WebCore::SharedBuffer>&&);
+    ~Session() override;
+
+    bool isValid() const { return m_session.get() && m_message && !m_message->isEmpty(); }
+    const String& id() const { return m_id; }
+    Ref<SharedBuffer> message() const
+    {
+        ASSERT(m_message);
+        return Ref<SharedBuffer>(*m_message.get());
+    }
     bool needsIndividualization() const { return m_needsIndividualization; }
     const Ref<WebCore::SharedBuffer>& initData() const { return m_initData; }
-    std::pair<media::OpenCdm::KeyStatus, String> update(const uint8_t* data, unsigned length);
-    std::pair<bool, String> load();
-    std::pair<bool, String> remove();
-    bool close() { return !m_session.Close(); }
-    media::OpenCdm::KeyStatus lastStatus() const { return m_lastStatus; }
-    bool containsInitData(const InitData& initData) const {
+    void generateChallenge(ChallengeGeneratedCallback&&);
+    void update(const uint8_t*, unsigned, KeyUpdatedCallback&&);
+    void load(SessionChangedCallback&&);
+    void remove(SessionChangedCallback&&);
+    bool close() { return !opencdm_session_close(m_session.get()); }
+    OCDMKeyStatus status() const
+    {
+        // FIXME: This API requires keyID yet there doesn't seem to be a way to retrieve one.
+        // Moreover it ignores it anyway this is why we can get away with passing nullptr for now.
+        return m_session ? opencdm_session_status(m_session.get(), nullptr, 0) : StatusPending;
+    }
+    bool containsInitData(const InitData& initData) const
+    {
         return m_initData->size() >= initData.sizeInBytes() && memmem(m_initData->data(), m_initData->size(), initData.characters8(), initData.sizeInBytes());
     }
 
 private:
+    struct SessionDeleter {
+        OpenCDMError operator()(OpenCDMSession* ptr) const { return opencdm_destruct_session(ptr); }
+    };
+    using ScopedSession = std::unique_ptr<OpenCDMSession, SessionDeleter>;
+
     Session() = delete;
-    Session(const media::OpenCdm& source, Ref<WebCore::SharedBuffer>&& initData);
+    Session(OpenCDMAccessor&, const String&, const char*, Ref<WebCore::SharedBuffer>&&, LicenseType, Ref<WebCore::SharedBuffer>&&);
+    void challengeGeneratedCallback(RefPtr<SharedBuffer>) override;
+    void keyUpdatedCallback(RefPtr<SharedBuffer>) override;
+    void messageReceivedCallback(RefPtr<SharedBuffer>) override;
+    void loadFailure() { keyUpdatedCallback(nullptr); }
+    void removeFailure() { keyUpdatedCallback(nullptr); }
```

Comment:
`nullptr` looks good but even better looks nothing. I'd try one of the following Either in the Observer or in the override, try to create a default value for the buffer being `nullptr`. If this can't be done because the compiler goes bananas somehow, I think we should overload it here.